masonry/src/passes/event.rs

```diff
@@ -0,0 +1,277 @@
+) -> Handled {
+    let _span = info_span!("text_event").entered();
+    if !event.is_high_density() {
+        debug!("Running ON_TEXT_EVENT pass with {}", event.short_name());
+    }
+
+    let target = root.state.focused_widget;
+
+    let mut handled = run_event_pass(root, root_state, target, event, |widget, ctx, event| {
+        widget.on_text_event(ctx, event);
+    });
+
+    // Handle Tab focus
+    if let TextEvent::KeyboardKey(key, mods) = event {
+        if handled == Handled::No && key.physical_key == PhysicalKey::Code(KeyCode::Tab) {
+            if !mods.shift_key() {
+                root.state.next_focused_widget = root.widget_from_focus_chain(true);
+            } else {
+                root.state.next_focused_widget = root.widget_from_focus_chain(false);
+            }
+            handled = Handled::Yes;
+        }
+    }
+
+    if !event.is_high_density() {
+        debug!(
+            focused_widget = root.state.focused_widget.map(|id| id.0),
+            handled = handled.is_handled(),
+            "ON_TEXT_EVENT finished",
+        );
+    }
+
+    handled
+}
+
+pub fn root_on_access_event(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    event: &AccessEvent,
+) -> Handled {
+    let _span = info_span!("access_event").entered();
+    debug!("Running ON_ACCESS_EVENT pass with {}", event.short_name());
+
+    let target = Some(event.target);
+
+    let handled = run_event_pass(root, root_state, target, event, |widget, ctx, event| {
+        widget.on_access_event(ctx, event);
+    });
+
+    debug!(
+        focused_widget = root.state.focused_widget.map(|id| id.0),
+        handled = handled.is_handled(),
+        "ON_ACCESS_EVENT finished",
+    );
+
+    handled
+}
+
+// ---
+
+fn run_event_pass<E>(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    target: Option<WidgetId>,
+    event: &E,
+    pass_fn: impl FnMut(&mut dyn Widget, &mut EventCtx, &E),
+) -> Handled {
+    let mut pass_fn = pass_fn;
+
+    let mut target_widget_id = target;
+    let mut is_handled = false;
+    while let Some(widget_id) = target_widget_id {
+        let (widget, pass_ctx) = get_widget_mut(&mut root.widget_arena, widget_id);
+        let parent_id = pass_ctx.parent();
+
+        let mut ctx = EventCtx {
+            global_state: &mut root.state,
+            widget_state: pass_ctx.widget_state.item,
+            widget_state_children: pass_ctx.widget_state.children,
+            widget_children: pass_ctx.widget_children,
+            is_handled: false,
+            request_pan_to_child: None,
+        };
+
+        if !is_handled {
+            trace!(
+                "Widget '{}' #{} visited",
+                widget.short_type_name(),
+                widget_id.to_raw(),
+            );
+
+            pass_fn(widget, &mut ctx, event);
+            is_handled = ctx.is_handled;
+        }
+
+        merge_state_up(&mut root.widget_arena, widget_id, root_state);
+        target_widget_id = parent_id;
```

Comment:
So unlike browser behavior, this doesn't do a capturing phase (yet?), only bubbling.

(Just noticing)

Reply:
We decided against a capturing phase, in the hope that it won't be necessary.